pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -0,0 +1,399 @@
+// service which is co-located with the sidecar proxy.
+func buildHTTPFilter(hostName model.Hostname, store model.IstioConfigStore) (*http_conn.HttpFilter, error) {
+	service := string(hostName)
+	split := strings.Split(service, ".")
+	if len(split) < 2 {
+		return nil, fmt.Errorf("failed to extract namespace from service: %s", service)
+	}
+	namespace := split[1]
+
+	roles := store.ServiceRoles(namespace)
+	if roles == nil {
+		return nil, fmt.Errorf("failed to get ServiceRoles in namespace %s", namespace)
+	}
+
+	bindings := store.ServiceRoleBindings(namespace)
+	if bindings == nil {
+		return nil, fmt.Errorf("failed to get ServiceRoleBinding in namespace %s", namespace)
+	}
+
+	log.Debugf("%s: converting RBAC rules to proxy config", RbacFilterName)
+	config, err := convertRbacRulesToFilterConfig(service, roles, bindings)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert RBAC rules to filter config: %v", err)
+	}
+
+	return &http_conn.HttpFilter{
+		Name:   RbacFilterName,
+		Config: util.MessageToStruct(config),
+	}, nil
+}
+
+// convertRbacRulesToFilterConfig converts the current RBAC rules (ServiceRole and ServiceRoleBindings)
+// in service mesh to the corresponding proxy config for the specified service. The generated proxy config
+// will be consumed by envoy RBAC filter to enforce access control on the specified service.
+func convertRbacRulesToFilterConfig(service string, roles []model.Config, bindings []model.Config) (*rbacconfig.RBAC, error) {
+	// roleToBinding maps ServiceRole name to a list of ServiceRoleBindings.
+	roleToBinding := map[string][]*rbacproto.ServiceRoleBinding{}
+	for _, binding := range bindings {
+		bindingProto := binding.Spec.(*rbacproto.ServiceRoleBinding)
+		roleName := bindingProto.RoleRef.Name
+		if roleName == "" {
+			log.Errorf("ignored invalid binding with empty RoleRef.Name: %v", *bindingProto)
+			continue
+		}
+		roleToBinding[roleName] = append(roleToBinding[roleName], bindingProto)
+	}
+
+	rbac := &policyproto.RBAC{
+		Action:   policyproto.RBAC_ALLOW,
+		Policies: map[string]*policyproto.Policy{},
+	}
+
+	for _, role := range roles {
+		// Constructs the policy for each ServiceRole.
+		var policy *policyproto.Policy
+		principals := convertToPrincipals(roleToBinding[role.Name])
+		log.Debugf("checking role %v for service %v", role.Name, service)
+		for i, rule := range role.Spec.(*rbacproto.ServiceRole).Rules {
+			if stringMatch(service, rule.Services) {
+				log.Debugf("role %v (access rule index %d) matched", role.Name, i)
+				if policy == nil {
+					policy = &policyproto.Policy{
+						Permissions: []*policyproto.Permission{},
+						Principals:  []*policyproto.Principal{},
+					}
+				}
+				// Generates the policy if the service is matched to the services specified in ServiceRole.
+				policy.Permissions = append(policy.Permissions, convertToPermission(rule))
+				policy.Principals = principals
```

Comment:
why not append? If this is want you want, you can set it at the time create policy object (line 200)

Reply:
The `principals` is already appended in `convertToPrincipals()`.